image-upload-with-preview/README.md

```diff
@@ -0,0 +1,289 @@
+    return unless @length
+    new FileRemove(this)
+
+) jQuery
+```
+
+```coffee
+# app/assets/javascripts/users/edit-profile.coffee
+
+$(document).on "ready page:load", ->
+  $("[data-file-remove]").fileRemove()
+  $("[data-file-preview]").filePreview()
+```
+
+```ruby
+# app/models/user.rb
+
+class User < ActiveRecord::Base
+  # ...
+
+  attachment :avatar
+
+  # ...
+end
+```
+
+```ruby
+# app/sanitizers/user/parameter_sanitizer.rb
+
+  def account_update
+    default_params.permit(USER_PARAMS, :current_password, :avatar, :remove_avatar)
+  end
+```
+
+```slim
+/ app/views/users/registrations/edit.html.slim
+
+.edit-profile-photo
+  .wrap
+    = attachment_image_tag(resource, :avatar, :fill, 200, 200, format: "jpg",
+        fallback: "defaultavatar.png", id: "avatar-image")
+  .actions
+    .change
+      = f.input :avatar, as: :attachment, direct: true, presigned: true
+          input_html: { data: { "file-preview" => { selector: "avatar-image",
+            remove_selector: "[data-file-remove]" } } }
+
+    .remove
+      = link_to "Remove avatar", "#",
+          data: { "file-remove" => { selector: "avatar-image", hidden: "user_remove_avatar",
+            file_input: "user_avatar", default: image_path("defaultavatar.png") } },
+          class: "button small alert #{resource.avatar.nil? ? 'hidden' : ''}"
+      = f.check_box :remove_avatar, class: "hidden"
+```
+
+Unfortunately `Refile::FileDouble` does not work with capybara.
+Before running tests save any lightweight image as `spec/fixtures/avatar.png`.
+For instance [this one](http://i7.5cm.ru/i/TWFB.png).
+
+```ruby
+# spec/support/features/shared_contexts/upload_avatar.rb
+
+shared_context "upload avatar" do
+  let(:file) { File.open(Rails.root.join("spec/fixtures/avatar.png")) }
+
+  def visit_profile_settings
+    visit edit_user_registration_path(current_user)
+  end
+
+  def upload_avatar
+    visit_profile_settings
+    fill_form(:user, avatar: file)
+    click_on "Update"
+  end
+
+  background do
+    stub_request(:any, /amazonaws.com/)
+  end
+end
+```
+
+```ruby
+# spec/features/user/account/upload_avatar_spec.rb
+
+require "rails_helper"
+
+feature "Upload Avatar" do
+  include_context "current user signed in"
+  include_context "upload avatar"
+
+  scenario "User uploads avatar" do
+    upload_avatar
+    visit_profile_settings
+
+    expect(page).to have_css("img[alt='Avatar']")
+  end
+end
+```
+
+```ruby
+# spec/features/user/account/remove_avatar_spec.rb
+
+require "rails_helper"
+
+feature "Remove Avatar", js: true do
+  include_context "current user signed in"
+  include_context "upload avatar"
+
+  background do
+    upload_avatar
+    visit_profile_settings
+  end
+
+  scenario "User uploads avatar" do
+    click_on "Remove avatar"
+    click_on "Update"
+
+    visit_profile_settings
+
+    expect(page).to have_css("img[alt='Defaultavatar']")
+  end
+end
+```
+
+Allow to connect to codeclimate from tests so that build would not brake up on semaphore.
```

Comment:
`Allow to connect to CodeClimate from tests so that build would not break up on SemaphoreCI`